### ISSUE TITLE
iOS: Notification Center Widgets for feed/folder

### ISSUE THREAD

**Issue by samuelclay** [samuelclay/NewsBlur#1162]
Just the top 4/5 stories from a selected feed or folder. Pretty straight-forward and I think this will be a real nice feature.

**Comment by samuelclay** [comment]
Yeah this will require coordination on my end which I can't give until May. We'll want to have a single endpoint that gives back 4 stories with the icon and photo data embedded. Let's hold off until then. This shouldn't take me too long at all to do on my end but we'll want to think about anything else we'll need. Can we pass cookies or should we also include the `secret_token` so we don't need the normal authentication?

**Comment by Dejal** [comment]
Yeah, one request that returns everything needed would be great. I think including a `secret_token` like the sharing extension would be the best solution.

**Comment by samuelclay** [comment]
There is not a day that goes by that I don't want this to exist. I've got widget fever (@caleb-allen is also about the launch widgets for android).

**Comment by samuelclay** [comment]
Default to All Site Stories and allow the user to use a feed/folder chooser in a dialog from the feed list menu.

**Comment by Dejal** [comment]
I'm going to have a go at this in the next day or two.  I currently plan to use the `/reader/refresh_feed/` endpoint to get the stories for a feed; I think that is the most sensible one, though maybe you can advise otherwise.  (Maybe `/reader/read_stories/` etc would be better?)

My only question at this point is about authentication.  I gather auth is handled via cookies, so presumably the extension will be able to leverage that.  If not, we may need a `secret_token` like the sharing extension.

I assume you won't have time to immediately whip up a new endpoint, despite your widget fever, so I'll try using the above one with cookies, and let you know if that doesn't work. We can always swap in another one later.

**Comment by samuelclay** [comment]
I'm done with classes on the 9th of December, so I can get to any endpoint requests then. Although my memory may be rusty but wouldn't `/reader/refresh_feed` force an unread recount? I would think you'd want to combine http://www.newsblur.com/api#/reader/unread_story_hashes and http://www.newsblur.com/api#/reader/river_stories

**Comment by samuelclay** [comment]
Actually refresh_feed is probably the insta-fetch stories endpoint, which would be a slow way of retrieving stories.

**Comment by Dejal** [comment]
Yeah, it is the insta-fetch one... which is indeed not so insta. It looked like a reasonable candidate, since it has all the info needed, though perhaps too much info. I'll look into combining `/reader/unread_story_hashes` and `/reader/river_stories`.

**Comment by samuelclay** [comment]
Actually, just hit /reader/load_river_stories and set the limit to 5. Pass in the feeds (or leave empty for All Site Stories). That way it's just a single call. No need to grab all the story hashes.

**Comment by Dejal** [comment]
Yeah, I was just looking at the [API docs](http://newsblur.com/api#/reader/river_stories), and thought that the `feeds` variation of `/reader/river_stories` (instead of the `h` one) looked like a good option for a single call, since it can return the newest unread stories. I don't see a `limit` parameter (maybe it's undocumented, like the `global_feed` one), but can just get the first page (if not all hidden). Simple enough.

**Comment by Dejal** [comment]
Ah yes, looking at the endpoint's source, I see there is a hidden `limit` parameter.  Goodo, that'll help.

**Comment by Dejal** [comment]
I now have a working widget that loads and caches stories using non-authenticated feeds.  I'm out of time, and it isn't finished, so I won't commit it yet.  I'll finish it in the second half of December, if you're able to add an endpoint by then.

The authentication cookies don't carry over to the widget extension.  It is actually possible to share cookies with extensions, but that would require logging in again to set the shared cookies, which isn't ideal.

So please do make a new endpoint that takes a token.  Probably the easiest option would be to just modify the existing `/reader/river_stories` to add an optional token parameter, or wrap that with a new endpoint.

If you wanted to customize it further, you could add an endpoint that takes a token and a list of feed IDs, and returns the 5 newest unread stories for those feeds.  For each story, I only need these properties:

- `story_hash`
- `story_feed_id`
- `short_parsed_date`
- `story_authors`
- `story_title`
- `story_content`
- `image_urls`

But again, just using a variation of the existing `/reader/river_stories` endpoint would suffice. I can ignore unneeded properties.

**Comment by Dejal** [comment]
I'll be ready to resume work on this in a couple days, if you're able to add the endpoint by then. Want a widget for the holidays? (Otherwise I'll work on other stuff, but it'd be nice to finish this.)

**Comment by samuelclay** [comment]
Ok, you got it. Every endpoint should now work with `secret_token` passed in either as a GET or POST parameter.

**Comment by Dejal** [comment]
Done!

### Widget:

- Added a **NewsBlur Latest** widget (scroll to the bottom of the widgets and edit as usual to show it).
- Displays the latest story in compact size, or the 5 latest in expanded size.
- The stories come from feeds chosen in the new **Widget Sites** view in the app (see below).
- Each cell is laid out like the default layout in the iOS app, with feed icon and name, a couple of lines each of title and content, author, time, and thumbnail image.
- The stories and images are fetched and cached for faster loading.
- Tap a cell to display that story in the app.
- If there are no sites chosen, or no new stories for those sites, appropriate messages are displayed; tapping the message opens the **Widget Sites** view in the app.
- The widget was written in pure Swift.

### App:

- In the iOS app, added a **Widget Sites** item to the main menu (I used a copy of one of the existing icons; you may want to substitute a different one).
- Choosing that displays a feed chooser view (a variation of the same one that is used for **Mute Sites** and **Organize Sites**).
- Any number of feeds can be selected (tap a folder to toggle all within), with the same **Select All** etc commands in the action menu as those other views.
- By default, all feeds are selected.  The selections are saved on a per-device basis (in the group container, so the widget can access them).

I'm really happy with how this has turned out. I hope you are too!

**Comment by samuelclay** [comment]
Wow this is incredible and works great. Here's a couple issues:

* When I first enabled it, it just showed "On its way..." and didn't change. I then tapped on that and returned to the widget screen and it worked. Can you somehow reproduce this? It'll be hard and might require you to delete and reinstall the app in order to get back into this condition, but it's the first experience and I want it to work for that first-time without having to tap first. And if we do have to tap first, just make a button that says that, "Tap to turn on".
* I realize that your system of choosing individual sites makes this a bit harder to do, but if they tap a story, I'd like it to open in the All Site Stories river view. Problem is that by selecting sites, the story goes further down the first page, but I'm ok with that because eventually it'll be found (since only latest unread stories show up in the river view). I'd rather the next/previous stories experience matches the widget listing.

**Comment by Dejal** [comment]
The "On its way..." occurs as there is no cached data initially, so nothing to display until the first fetch completes. Plus I had difficulty in getting it to reload once the data is received. I'll have another go at improving that.

If I can't, I could reword it to something like "Tap to configure in NewsBlur", which would display the Widget Sites view in the app (as it does now). I've seen other widgets do that on first use.

Should be feasible to open in the All Stories river view. As you say, it may take longer to display it, but probably okay; I'll try that.

**Comment by samuelclay** [comment]
Instead of having the user configure it, just automatically select all sites. The app is already doing that once the widget dialog is opened. I believe you can pass in no sites at all and get everything.

**Comment by Dejal** [comment]
Yes, all sites are selected by default. The issue is just what to display before the widget has loaded anything. Even if I leave the initial display with a message, if the user ignores that and comes back to the widget later, it'll be populated. So a message to configure it is just a placebo (though also useful for discovery of the preferences; people might want to trim down what to show). But as I said, I'll see if I can get it to automatically redisplay once loaded.

**Comment by nriley** [comment]
This is a neat idea. I saw a few issues on iPad with this version.

1. The first time I tapped "Please choose sites to show" (this should probably look more like a button to indicate that you can tap on it) the site selector appeared but immediately closed - looks like it may have been preempted by the stories refreshing in the background?  The second time it worked

2. I can't get anything to display even after seleting sites. I suspect that communication between the widget and app aren't quite working right but not sure how to troubleshoot (this may have to do with how I have had to make changes to use my own bundle IDs etc. to get it to build).  Of note when I press and hold on the NewsBlur icon the widget does not appear even though it does on the home screen:

<img width="1112" alt="image" src="https://user-images.githubusercontent.com/68303/71312585-96fa3e80-23fa-11ea-9396-017cec97a66d.png">

**Comment by samuelclay** [comment]
Does a user have to tap on a widget to make it work? I've seen this with a few others, so is it required that the user opens the widget so that cookies/auth can be used?

**Comment by samuelclay** [comment]
> I have had to make changes to use my own bundle IDs etc. to get it to build

Are you changing bundles so notifications continue to work with the original app? Otherwise, if you're overwriting the App Store installation you shouldn't need to change bundle IDs. @Dejal I noticed you have a dejal-specific target. Somehow related?

**Comment by nriley** [comment]
I've been building it as a separate app for years (you can see two NewsBlur icons in the screenshot above).  [Here](https://github.com/samuelclay/NewsBlur/files/3991738/njrBlur.txt) is the patch I'm using.

I get lots of Xcode signing errors if I try to build as is so I have to change to my own development team and application group. This has never been a problem before. Though I doubt it is related, I also had to switch Xcode 11.3 back to the legacy build system.

**Comment by Dejal** [comment]
I have a Dejal-specific target for Turn Touch, not NewsBlur.  As I recall, we couldn't add me to Turn Touch as it is an individual developer account, instead of a company one. I expect other developers would have the same problem with NewsBlur too.

Anyway, @nriley, changing the app group might break communication with the widget. Make sure the app and widget both have the same app group. The group is also hardwired in the code, so you'd have to change both the project (app and widget entitlements) and the app and widget code.  Search for `group.com.newsblur.NewsBlur-Group`; it occurs 9 times. Though it may not be necessary to change it at all.

The widget doesn't have to be tapped to work.

**Comment by Dejal** [comment]
Sam: I notice that you renamed the widget from "NewsBlur Latest" to just "NewsBlur". That doesn't compile; the widget can't have the same name as the the app. I actually initially just called it "NewsBlur " with a space on the end to avoid that issue, but didn't like that, so changed it to "NewsBlur Stories", then to "NewsBlur Latest". You're welcome to choose a different name, but it has to be unique. Let me which you'd prefer.

**Comment by nriley** [comment]
@Dejal, thanks; did not think about the app group being in the code. I could only find 6 occurrences but I changed them all. Now I got "On my way…" in the widget but I had to tap on it, then re-enter the app to select feeds again.

I was able to get a single article to display in the collapsed mode of the widget but when I tried to expand it, the widget displayed "Unable to Load". Crash log is [attached].(https://github.com/samuelclay/NewsBlur/files/3991751/NewsBlur.12-21-2019.2-47.PM.txt)

**Comment by Dejal** [comment]
Interesting... I hadn't seen that crash before. See if you can reproduce it. If so, put a breakpoint in `viewWillTransition(to:with:)` and see what the values are. To debug it, you'll have to use Debug ▸ Attach to Process by PID or Name...

**Comment by nriley** [comment]
Not sure if this is what you want — we could set up a screen sharing session if helpful.

![image](https://user-images.githubusercontent.com/68303/71313126-76ce7d80-2402-11ea-9416-20161bf64256.png)

**Comment by nriley** [comment]
I was able to get the widget to expand successfully now but it is still crashing before I can tap on anything due to running out of memory.

**Comment by Dejal** [comment]
Widgets are very memory-constrained; I've seen it get a memory warning at 22 MB used, and killed at 25 MB used. I don't have that issue on my devices (original iPad Pro 12", and iPhone 11 Pro Max), but if you have an older one, I could see that being a problem, depending on the images being loaded.

I scale the story images down to the thumbnail size before caching them. They are cached in memory and disk, but the memory cache is flushed on a memory warning.

You could try disabling the memory cache, to see if that helps for you. Comment out the two instances of `imageCache[identifier] = image`. (That might have a small performance hit, having to load from disk each time, but probably negligible.)

@samuelclay could help by providing a pre-scaled thumbnail endpoint. Similar to the `/reader/favicons` one, an endpoint that takes a story hash and desired size, and returns the first image (if any) scaled to the requested size (square). That'd avoid having to fetch a potentially huge image from the story.

**Comment by nriley** [comment]
I'm using a 2nd gen 10.5ʺ iPad Pro, so it should have the same amount of memory as your 12ʺ.

FYI, commenting out the imageCache lines does not seem to help. In the very short time when the widget draws before being killed, I do notice that the images appear to be being stretched out into a square. I've [attached](https://sabi.net/temp/widget.mp4) a short video of what happens when the widget tries to display.

**Comment by Dejal** [comment]
Good point about the aspect ratio; I'll fix that.

Another idea: try disabling the thumbnail images entirely, to confirm that is the issue for you. Comment out these lines in the widget controller:

        storyImage(for: story.id, imageURL: story.imageURL) { (image, id) in
            if story.id == id {
                cell.thumbnailImageView.image = image
            }
        }

If it works with no thumbnail images, that suggests the thumbnail endpoint would be worthwhile. If it doesn't, we may have to investigate further; could be too much story data or something.

**Comment by nriley** [comment]
Disabling the thumbnail images entirely worked; it doesn't crash any more.

Another display issue now I can get the display to stick around for a bit — it doesn't look like you do anything with long story titles, they're just cut off.  A smaller font and/or ellipsis would be helpful.

**Comment by Dejal** [comment]
Okay, @samuelclay, can you please add that endpoint to take a story hash and desired size, and return the first image (if any) scaled to the requested size. Either use the smaller dimension as the size (and I'll aspect fill to cut off the extra), or have the server do that and return as a square image.

**Comment by samuelclay** [comment]
Two visual items:
* Like @nriley suggested, can we add ellipses for cut-off story titles
* The font face, size, and colors between the widget and the app are different. See these screenshots. I'd love to see the same fonts and styles from the app in the widget. Including the colors bars on the left of the story title.

![IMG_7152](https://user-images.githubusercontent.com/44229/71372156-8b854f80-2581-11ea-8be6-354853d67fbd.PNG)
![IMG_7153](https://user-images.githubusercontent.com/44229/71372159-8de7a980-2581-11ea-96c0-38c68ed9e80c.PNG)

**Comment by samuelclay** [comment]
Well there's the secure image proxy that I just launched, but it doesn't do resizing and cropping. I suppose I can add options to it but that's non-trivial. I think it'd be easier to figure out why images are crashing the widget. I realize that we want the widget to be as fast as possible, so building thumbnails into the image proxy wouldn't be a terrible idea, but that's going to take me some time and I'm not sure I'm going to find that time due to my thesis obligations.

**Comment by samuelclay** [comment]
https://github.com/willnorris/imageproxy looks promising as a replacement to camo, so I'll start there and try to get that deployed

**Comment by Dejal** [comment]
- Adding an ellipsis is non-trivial; I'd need to calculate the bounds of the text (given the content and font), and truncate the string to that length. In the app, that is done as part of drawing. But sure, it'd be tidier, and certainly doable.
- I'll see if I can make the fonts etc match the app more closely, and include the color bars.
- If you can add an image endpoint, that should help. If you can't, I could try to limit memory usage some more. The issue is that widgets have a strict memory cap, and a story image could be arbitrarily large, so any story image could push it over the limit. I could perhaps omit excessively large images if necessary.

**Comment by nriley** [comment]
> * Adding an ellipsis is non-trivial; I'd need to calculate the bounds of the text (given the content and font), and truncate the string to that length. In the app, that is done as part of drawing. But sure, it'd be tidier, and certainly doable.

Unless I'm missing something this should be built into UIKit — I was able to do this just by editing the xib. This is less of an issue if the title is 2 lines but if it's just one I think it is necessary.

IB configuration:
<img width="331" alt="image" src="https://user-images.githubusercontent.com/68303/71379765-97314000-259a-11ea-884b-9e83db478c8e.png">

Output:
<img width="376" alt="image" src="https://user-images.githubusercontent.com/68303/71379774-a3b59880-259a-11ea-8bf1-a6fe4ef28a2b.png">

**Comment by Dejal** [comment]
Yeah, that only works if the label is 1 line. Not compatible with word wrap for two lines.

**Comment by nriley** [comment]
I was only seeing one line on the iPad, maybe because it was picking a
bigger font size.  If you can get 2 lines to display on all platforms
that'd work, otherwise I'd prefer 1 line with an ellipsis to 1 line without!

**Comment by Dejal** [comment]
I'm out of time for the month, but I think this is done. I'll leave this open for now in case anything else comes up, but unless super urgent, it'll have to wait for next month.

Happy holidays!

**Comment by Dejal** [comment]
I notice that the thumbnails sometimes show blue "Err" text instead of the image (or sometimes "Tw"); I assume the thumbnailer can't load the image or something. You might want to investigate that, and perhaps omit the image if it can't be loaded.


![IMG_8066](https://user-images.githubusercontent.com/1318849/73033339-90deef80-3df6-11ea-93cf-b9131f380df6.png)


Also, the ideal size for the thumbnails would be 192 (being 64 pixels at 3x).  But 200 is close enough.

**Comment by samuelclay** [comment]
That's "Email" and "Twitter" from the image buttons on the bottom of Feedburner feed stories.

Just changed the signed url to use 192px.

**Comment by samuelclay** [comment]
Looking amazing! Ok, here's a couple small issues: 

* Every title is a single line. Can we bump this to 2, or better yet, use the preference?
* Any chance we can remove the space when there's no image? (Is it possible there is an image but nothing is coming up in the image proxy? I can investigate, although it's a bit of a blackbox)
* The fifth story is cut off. This might get complicated with the story title height preference.

![IMG_7388](https://user-images.githubusercontent.com/44229/73038387-3d819700-3e20-11ea-9794-1ebe77f793e1.PNG)

**Comment by Dejal** [comment]
Hmm, I'm not sure why the titles are one line for you.  As you can see in the screenshot I included in my previous comment, they are two lines for me.  Might be something to do with the screen metrics or system font preferences.

I do leave the thumbnail space empty, even when there isn't an image.  I should be able to adjust the auto-layout constraints to avoid that.

The last row being too short might be a clue on what is happening.  Or may be a symptom of using a small screen, which could impose a smaller maximum height.  Which device are you using?

**Comment by Dejal** [comment]
Never mind; an iPhone 8 in the Simulator appears like your screenshot. I'll see what I can do to improve it.

**Comment by samuelclay** [comment]
Yep, I'm on an iPhone 7 (and will continue to be until the iPhone 12 comes out in the Fall)

**Comment by samuelclay** [comment]
So great! No issues that I can find after 5 minutes of use, so that's a great sign. I'll be playing around with it constantly, but I think we're on our way to launching soon.
